### PR TITLE
refactor: Better source code organization

### PR DESCRIPTION
**Pull request by tibetiroka**
**Refactor**

This PR addresses a long-standing issue with the organization of the source code

## Summary
Sorts existing files into various directories, somewhat clearing the mess that was the source directory.

As usual, the division and naming of directories is up for discussion.

## Testing Done
It builds.

### CODE REVIEW

**Comment by petervdmeer** [comment]
> As usual, the division and naming of directories is up for discussion.

Here is my contribution to this discussion: https://github.com/endless-sky/endless-sky/pull/7128

**Comment by tibetiroka** [comment]
> Here is my contribution to this discussion: #7128

Huh. That's a radically different take, working with independent modules instead of just directories. It's certainly interesting, though it still dumps most of the files into the runtime directory. Perhaps we could combine the two solutions somehow?

We still support scons, so maybe it's not time for that change just yet?

**Comment by Koranir** [comment]
I prefer this PR over 7128, to be honest. It's better for the developer experience, and more organized - the two PRs have different goals imo, and separating libs isn't _that_ high of a priority yet.

**Comment by petervdmeer** [comment]
> Perhaps we could combine the two solutions somehow?

Yes, that is what I was hoping for.

I'm not asking you to take on the library split into this PR and I'm not looking for putting everything into a single runtime directory, but let's try not to put files that would go into different libraries into the same directory.

We will not be able to get a complete 100% correct split over libraries today; it is very likely that we will still need to move some files/code during refactoring, but awareness of this idea of splitting is probably already beneficial for this PR.

> We still support scons, so maybe it's not time for that change just yet?

Supporting scons or not is not very relevant for splitting into multiple libraries. If we want to split the codebase into multiple libraries, then we just need to do that for all the build-systems we support. Supporting less build systems obviously makes such change easier, but scons is not blocking.

The main blocker for splitting the codebase into multiple libraries is that most people are not yet aware of the benefits of such a split.

**Comment by petervdmeer** [comment]
> I prefer this PR over 7128, to be honest.

That's fine, but I wasn't offering #7128 as an alternative. I was also hoping to combine the two (as tibetiroka suggested).

> It's better for the developer experience, and more organized

If we combine them, yes. If we put files together in directories, and then compile files from the same directory into different libraries, then that would probably introduce a lot of frustration.

> - the two PRs have different goals imo

I disagree. Both this PR and the split into libraries seem to be working towards improving the codebase and improving our development processes. 

> and separating libs isn't _that_ high of a priority yet.

The LTO (Link Time Optimization) problem actually prevented some feature-PRs from getting merged (or at least significantly blocks the ability to provide unit-test-coverage for them) and is quite high on the list of tehhowch, quyykk and me of problems that are hurting the development of ES. I feel it should get addressed with more priority than that it currently gets.

**Comment by Koranir** [comment]
> The LTO (Link Time Optimization) problem actually prevented some feature-PRs from getting merged (or at least significantly blocks the ability to provide unit-test-coverage for them) and is quite high on the list of tehhowch, quyykk and me of problems that are hurting the development of ES. I feel it should get addressed with more priority than that it currently gets.

What's the issue with LTO? Honest question, never encountered a problem with it before - correct me if I'm wrong, but it should only be stripping unused symbols which should be fine.

**Comment by petervdmeer** [comment]
> What's the issue with LTO? Honest question, never encountered a problem with it before - correct me if I'm wrong, but it should only be stripping unused symbols which should be fine.

Yes LTO is doing exactly that. What happened is that we use LTO for the unit-tests to strip out symbols that are unused _during the unit tests_. This trick allowed us to avoid including external libraries like libpng, while still testing classes that depends on such libraries.

So LTO is not a problem by itself, the problem is that we used it as a somewhat dirty workaround to test classes in our unit-tests while the dependencies are not full-filled.

This dirty workaround doesn't always work.

File: source/CMakeLists.txt [hunk 1]
```diff
@@ -14,358 +14,259 @@ endif()
...
+	image/ImageBuffer.cpp
+	image/ImageBuffer.h
+	image/ImageSet.cpp
+	image/ImageSet.h
+	image/RenderBuffer.cpp
+	image/RenderBuffer.h
+	image/Sprite.cpp
+	image/Sprite.h
+	image/SpriteSet.cpp
+	image/SpriteSet.h
```
Comment (petervdmeer) [inline source/CMakeLists.txt:210]: Putting images in their own directory does make sense. Maybe put this into a separate PR as well?

File: source/CMakeLists.txt [hunk 2]
```diff
@@ -383,4 +284,103 @@ target_sources(EndlessSkyLib PRIVATE
 	text/alignment.hpp
 	text/layout.hpp
 	text/truncate.hpp
+	ui/BankPanel.cpp
+	ui/BankPanel.h
```
Comment (petervdmeer) [inline source/CMakeLists.txt:288]: A lot of stuff is in the ui directory; I feel that panels and realtime-stuff like AmmoDisplay and Radar are better of into separate directories.

File: source/data/CategoryList.h [hunk 3]
```diff
@@ -13,8 +13,8 @@ You should have received a copy of the GNU General Public License along with
 this program. If not, see <https://www.gnu.org/licenses/>.
 */
 
-#ifndef CATEGORY_LIST_H_
-#define CATEGORY_LIST_H_
+#ifndef DATA_CATEGORY_LIST_H_
+#define DATA_CATEGORY_LIST_H_
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply source/data/CategoryList.h:17]: > In the [C](https://en.wikipedia.org/wiki/C_(programming_language)) and [C++](https://en.wikipedia.org/wiki/C%2B%2B) programming languages, #pragma once is a non-standard but widely supported [preprocessor directive](https://en.wikipedia.org/wiki/C_preprocessor) designed to cause the current header file to be included only once in a single compilation.

Since it's non-standard, do we really want to use this?

I mean, as far as I know, both GCC and Clang support it, and they seem to be using it extensively within their own libraries. From what I've heard, MSVC also supports it, though I can't test that. I'm also not filled with confidence by Wikipedia's "its implementation is tricky and might not always be reliable" comment.

File: source/CMakeLists.txt [hunk 4]
```diff
@@ -14,358 +14,259 @@ endif()
...
+	audio/Sound.cpp
+	audio/Sound.h
+	body/Armament.cpp
+	body/Armament.h
+	body/AsteroidField.cpp
+	body/AsteroidField.h
+	body/Body.cpp
+	body/Body.h
+	body/CaptureOdds.cpp
+	body/CaptureOdds.h
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply source/CMakeLists.txt:53]: Actually, there is no ship directory in this PR. Should there be one, and if so, should it be within `body`?

File: source/CMakeLists.txt (hunk 2 above)
Comment (Amazinite) [reply source/CMakeLists.txt:288]: `/ui/hud/`?

File: source/data/CategoryList.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/data/CategoryList.h:17]: This is now solved in the latest master-branch.

File: source/Logger.h [hunk 5]
```diff
@@ -13,8 +13,8 @@ You should have received a copy of the GNU General Public License along with
 this program. If not, see <https://www.gnu.org/licenses/>.
 */
 
-#ifndef ES_LOGGER_H_
-#define ES_LOGGER_H_
+#ifndef LOGGER_H_
+#define LOGGER_H_
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/Logger.h:17]: this is now solved in the latest master-branch (by pragma once)

File: source/CMakeLists.txt [hunk 6]
```diff
@@ -14,358 +14,259 @@ endif()
...
+	system/Weather.h
+	system/Wormhole.cpp
+	system/Wormhole.h
+	system/WormholeStrategy.h
+	test/Test.cpp
+	test/Test.h
+	test/TestContext.cpp
+	test/TestContext.h
+	test/TestData.cpp
+	test/TestData.h
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/CMakeLists.txt:269]: Done, in latest master

File: source/CMakeLists.txt [hunk 7]
```diff
@@ -14,358 +14,259 @@ endif()
...
 	opengl.cpp
 	opengl.h
 	pi.h
 	shift.h
+	audio/Audio.cpp
+	audio/Audio.h
+	audio/Music.cpp
+	audio/Music.h
+	audio/Sound.cpp
+	audio/Sound.h
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/CMakeLists.txt:45]: Done in a follow-up PR.

**Comment by tibetiroka** [comment]
This was meant to be a place to discuss the changes, so now that most of these changes are in the game / in separate PRs, I don't see any reason to keep this open anymore. (I will still double-check with the discussions here.)
